### ISSUE TITLE
Stock quantity are not verified at the last step of the checkout

### ISSUE THREAD

**Issue by kpodemski** [PrestaShop/PrestaShop#10762]
**Describe the bug**

As the remaining quantity in stock is not checked at the last step of the checkout, you are able to arrive in the payment process and to pay a product that is out of stock even if it's not allowed.

**To Reproduce**

1. Set a product quantity to 2 in the BO
2. Add this product to cart, 2 units
3. Go to the last step of the checkout (payment), just before you click "Confirm with an obligation to pay"
4. Now, change quantity of the product to 1 in back-office
5. Click "Confirm with an obligation to pay" on last step of the checkout (payment)
6. You are now able to pay while you shouldn't

**Wanted behavior**
We should check the stock quantity at the last step of the checkout while clicking on "Order with obligation to pay".
If the product is out of stock and if order of out of stock product isn't allowed, then we should bring the user at the cart, display the following error message "The item _product name_ is no longer available in this quantity. You cannot proceed with your order until the quantity is adjusted." and disabled the "proceed to checkout" button like below:

![capture d'écran_3016](https://user-images.githubusercontent.com/13449658/113314755-eb35e300-930c-11eb-9ec1-cda32efe4eac.png)

Please note that this "stock quantity check" and this behavior is already present at the previous checkout steps (personal informations, addresses, shipping method) when you click on "Continue"

**Additionnal information**
PrestaShop version: 1.7.0.0 to 1.7.5.0
PHP version: N/A

**Comment by khouloudbelguith** [comment]
Hi @kpodemski,

Thanks for your report.
I manage to reproduce the issue with the last version 1.7.4.2 / branch 1.7.5.x.
This PR: #8752 do not check the quantity of products available in the last step.
@marionf what do you think?

Best regards, Khouloud

**Comment by marionf** [comment]
Hello @kpodemski 
Yes, we are aware of this and we talked about it with @alegout & @rGaillard  when we added stock verification at each step but unfortunately for some technical reasons, we can't add this verification at the last step
@rGaillard if you want specify the technical reasons

**Comment by marionf** [comment]
@rGaillard explained me this :
In any case, there will always be a step where you will not be able to check the stock, for example on the payment provider's page
Adding an ajax check at the click of the button may slow down the process for the merchant's customer

**Comment by kpodemski** [comment]
@marionf 

i agree, but... maybe it would be a good idea to give merchants a choice, at this moment it is considered as bug by many people and unfortunately causing a lot of issues

**Comment by colinegin** [comment]
➤ marionf commented:

Hello @kpodemski 
Yes, we are aware of this and we talked about it with @alegout & @rGaillard  when we added stock verification at each step but unfortunately for some technical reasons, we can't add this verification at the last step
@rGaillard if you want specify the technical reasons

**Comment by kpodemski** [comment]
hey @rGaillard could you tell me more about those technical issues?

**Comment by colinegin** [comment]
@PrestaShop/prestashop-contributors what do you guys think ?

**Comment by rGaillard** [comment]
At this step your customer can be on your payment gateway payment page, how do you handle this case? You just refuse the return of your payment gateway with no link between your shop and your payment gateway? This is a bad idea...

**Comment by kpodemski** [comment]
Hmmm, you won't be on payment gateway before clicking "Payment with an obligation to pay", right? We're talking about 1.7 only of course, the only issue i see could be with PayPal Express module and modules which skips this step...

**Comment by colinegin** [comment]
Here is a solution that Axome gave us : 

- when customers reach the payment step, re-check stocks again (already done on the classic theme)
- when customers proceed to the payment, bank modules should be in "authorization" mode only : this allows to validate if the purchase is possible or not (the customer has virtually proceeded to the payment on the bank or payment platform).
- When returning to PrestaShop (PaymentModule::validateOrder method), there should then be a check on the stock and proceed to the order capture and creation ONLY if the stock allows it.
- If the stock does not allow it, the customer is redirected to his cart with an error message "this product is no longer available".

@eternoendless @PrestaShop/prestashop-core-developers @kpodemski @rGaillard what do you think of this solution ?

Apparently this bug was already present in 1.6. Can someone confirm ?

**Comment by janekx** [comment]
yes 1,6,1,24 have this bug too any fix_

**Comment by colinegin** [comment]
ping @PrestaShop/prestashop-core-developers

**Comment by nsorosac** [comment]
I'll be happy to discuss this solution with you!

> Here is a solution that Axome gave us :
> 
> * when customers reach the payment step, re-check stocks again (already done on the classic theme)
> * when customers proceed to the payment, bank modules should be in "authorization" mode only : this allows to validate if the purchase is possible or not (the customer has virtually proceeded to the payment on the bank or payment platform).
> * When returning to PrestaShop (PaymentModule::validateOrder method), there should then be a check on the stock and proceed to the order capture and creation ONLY if the stock allows it.
> * If the stock does not allow it, the customer is redirected to his cart with an error message "this product is no longer available".

**Comment by janekx** [comment]
Our issue in closer look. We use ASM 

Customer make cart and put items in them in one day, but doesnt process cart, not order. Next day some items comes 0 and system not allow to put them in new cart. So the safety order 0 products is OK. After that the customer come back to own cart from day ago and order without any trouble. The 0 items are ordered and we see -1 pcs on those items. 
As far as i remeber this bug we meet much earlier then in 1.6.1.24. We use custom template, can be this issue of the template or core system?

**Comment by anfuca** [comment]
I'm suffering this issue last days... Last Friday, I've reached -10 in a product stock because of this issue. So 10 customers that are dissapointed... 10 we are so sorry emails... 10 refunds. 

What would be the best approach to solve this? I'm not talking about a technical solution. 
What would be, in theory, the best solution?

This answer from Tobi (Shopify) from 2007:
> One example of something we do behind the scenes is that items are “reserved” when a customer moves to the payment step. For 5 minutes the products in his/her cart are reserved for the buyer. This accommodates people who have to go hunting their house for their credit card whenever they want to buy online :)

**Comment by anfuca** [comment]
> Here is a solution that Axome gave us :
> 
> * when customers reach the payment step, re-check stocks again (already done on the classic theme)
> * when customers proceed to the payment, bank modules should be in "authorization" mode only : this allows to validate if the purchase is possible or not (the customer has virtually proceeded to the payment on the bank or payment platform).
> * When returning to PrestaShop (PaymentModule::validateOrder method), there should then be a check on the stock and proceed to the order capture and creation ONLY if the stock allows it.
> * If the stock does not allow it, the customer is redirected to his cart with an error message "this product is no longer available".
> 
> @eternoendless @PrestaShop/prestashop-core-developers @kpodemski @rGaillard what do you think of this solution ?
> 
> Apparently this bug was already present in 1.6. Can someone confirm ?

@colinegin when you say that Axome gave you that solution... Do you mean a working solution? Is already implemented in your site? If yes, I'm very interested, and I would like to talk with them.

This bug is a very serious headache for me... I'm having a lot of orders with negative stock.

**Comment by nsorosac** [comment]
Hi @anfuca 

Unfortunately, this is not a working solution, but rather a proof of concept. We had to think of a solution for this problem for one of our customers, and we ended up with this idea.

It still needs to be reviewed by prestashop core team because it has a lot of implications:

- The final stock check should be a feature you can toggle on or off in the admin panel, because all payment gateways can't comply with this out of the box
- There should be additional documentation on this feature in the admin panel
- We can imagine that compatible addons could have a specific tag / label to help merchants choose the right payment module

This being said, if you only use direct payment modules (card or paypal) that can be configured in "authorization only" mode, to have a working solution you would need to: 
- Override the PaymentModule::validateOrder method to check stocks before creating the order
- Handle the different use-cases (OK ? Error ? Other ?) : let the order creation proceed, or go back to cart summary
- Capture the payment (depends on payment gateway, either you can do this with their code, or with their API)

**Comment by ChineseNorris** [comment]
Would like to follow and support the idea of "holding" the 5 minutes reserved stock while processing the payment, and check the stock right before the payment step.

Another idea is to force payments addons to send a stock check and throw a payment error if the stock is not available anymore, would help with this concurrency issue a lot.

**Comment by mdrbx** [comment]
We developped a module with this in mind plus some "real time" capabilities that can be disable for avoiding high server workload :  
https://addons.prestashop.com/en/stock-supplier-management/21707-temporary-product-reservation-lonely-stock.html

It creates a separate database and mostly rely on updateQty hook.
Module is sold since more than 3 years now, fully functionnal and debugged. Feel free to ask questions about this if needed, you can also take a look at our FAQ here : https://www.malttt.com/faq-products/?lang=en

**Comment by nsorosac** [comment]
@MattLoyeD your module is a nice workaround, and we used it for one of our projects :smile: 

We had severe performance issues that we managed to fix in the module (database indexes, externalized CRON task, etc.), I would be happy to share these optimizations with you :wink:

**Comment by anfuca** [comment]
@nsorosac So so true. That module has a very bad and serious perfomance problems. I would like to know those optimizations!

**Comment by HartLarsson** [comment]
well the check of stock in the last step seems someone have found a way to do it: https://addons.prestashop.com/it/gestione-scorte-fornitori/48536-checking-stock-at-checkout-before-paying.html#specifications

why not implementing on Prestashop Core too?

**Comment by HartLarsson** [comment]
> We developped a module with this in mind plus some "real time" capabilities that can be disable for avoiding high server workload :
> https://addons.prestashop.com/en/stock-supplier-management/21707-temporary-product-reservation-lonely-stock.html
> 
> It creates a separate database and mostly rely on updateQty hook.
> Module is sold since more than 3 years now, fully functionnal and debugged. Feel free to ask questions about this if needed, you can also take a look at our FAQ here : https://www.malttt.com/faq-products/?lang=en

your module is nice but it remove all abandoned cart, and remarketing on abandone cart is a big money recovering.

**Comment by hibatallahAouadni** [comment]
Ping @PrestaShop/prestashop-core-developers friendly reminder :wink:

**Comment by matks** [comment]
Hi everybody,

I think we can close this issue.

😄 

I'm quite sure a large number of people will think I'm crazy when I say that. 😅 

_Please let me explain before shouting at me_ 😉 

The reason why I say we can close this issue is that _this issue has been actually a global ecommerce issue_ for years. It's not just PrestaShop. I'm quite sur WooCommerce or Magento can have it too.

It has been summarized it well: when 2 customers are next to each other, and there's only 1 item left, how can we make sure that only the very 1st customer to buy pay, and not the other? The 2 payments could be separated only for a nanosecond. _Since online payment are asynchronous by design I think it is impossible to solve this issue_.

This problem is more famous in another form: plane seats booking. You have a plane with 100 seats. Only 1 seat is not booked yet. 2 customers are looking to book the last plane. How do you know which of the customer will be the very first? Since people use a payment gateway that is asynchronous and external to PrestaShop there is no reliable way to prevent the 2nd customer to pay... and then find out the last seat has already been booked.

As far as I know, there is 2 standard and widely used possible workarounds for online businesses:

## 1. Over-selling

This strategy aims to ignore the issue and, when you have 2 people buying the last item, either refund the unlucky customer or produce more of the items.

AFAIK most plane companies do that. Either they refund the unlucky customers who booked a seat that was not available anymore _or they expect that some customers will not come when the day comes to fly_. If however all customers come to onboard the plane, they apologize and reschedule. They do this because it's less expensive.

## 2. Under-booking

This strategy aims to sell only X of your stock but you keep X+30 items in warehouse. The 30 extra items not being sold are there if the issue above appears. "Oh, you have paid but the very last item has already been sold? Dont worry, we said we were only selling 500 items but we have 530, so you'll get what you paid for."

Some plane companies do that. The plane has 50 seats but only 45 can be booked. 5 seats are kept as extra, to handle this usecase or for other needs like people with specific conditions.

<hr>

**Both of these strategies are not solved by the software but by the strategy followed by the stock administrator. So I dont think we need to keep this issue open.**


@ldeprestashop @eternoendless @kpodemski do you agree with me that we can close this issue?

**Comment by marionf** [comment]
@matks 

> when 2 customers are next to each other, and there's only 1 item left, how can we make sure that only the very 1st customer to buy pay, and not the other? 

The issue is not only when 2 customers are next to each other.
The main problem is the non-verification of the stock at the last step of the checkout.

If I have 2 products "A" in stock, add these 2 "A" products to cart, change the quantity to 1 for "A" in BO and click on "Confirm with an obligation to pay" on the last checkout step, I am able to pay while I should not. I should have at the last checkout step when I click on "Confirm with an obligation to pay" the info that there is only one product left and that the order cannot be validated. We do it already at the others checkout steps, we could do it also at the last step when the user clicks on "Confirm with an obligation to pay"

> Since people use a payment gateway that is asynchronous and external to PrestaShop there is no reliable way to prevent the 2nd customer to pay

Yes, we will not be able to check the stock at the payment step, but we can do it at the last checkout step to limit the risk of having orders with unavailable products

**Comment by nsorosac** [comment]
> Since people use a payment gateway that is asynchronous and external to PrestaShop there is no reliable way to prevent the 2nd customer to pay

We shared this approach to solve that specific problem:

- https://github.com/PrestaShop/PrestaShop/issues/10762#issuecomment-561720953
- https://github.com/PrestaShop/PrestaShop/issues/10762#issuecomment-621041635

**TL;DR** asynchronous and external does not mean you can't check quantities before `PaymentModule::validateOrder` is called. It's not an easy task, I can't argue that, but it's doable.

**Comment by matks** [comment]
@nsorosac 

https://github.com/PrestaShop/PrestaShop/issues/10762#issuecomment-561720953 if I get it correctly this introduces a new step. This step is "ask customer bank if he has the money to pay"

I dont think it solves the issue 🤔  .

> When returning to PrestaShop (PaymentModule::validateOrder method), there should then be a check on the stock and proceed to the order capture and creation ONLY if the stock allows it.

What if 2 customers come back from the payment gateway in the same millisecond? Stock is still 1. Each verification of the stock available returns `true` because, yes, there is still 1 item available. Then both http requests are handled by webserver and both customers are made to paid. And both customers paid while there was only 1 item left.

There could be a valid solution, though ⚠️ . It would be **to process only one payment at a time** from PS side. This means to block traditional http request handling (in parallel) OR to use a mutex system and only handle 1 payment request at a time, effectively making sure each customer who pays has a reserved item. **This** could work.

That being said I'm afraid the amount of changes and work to achieve that is enormous, and as explained it only works with some payment gateways, not all.

@marionf I was following up to https://github.com/PrestaShop/PrestaShop/issues/10762#issuecomment-538371505 because @PrestaShop/prestashop-core-developers have been asked to answer about it 😉  .

> At this step your customer can be on your payment gateway payment page, how do you handle this case? You just refuse the return of your payment gateway with no link between your shop and your payment gateway? This is a bad idea...

@marionf It seems two topics got mixed in this issue 😄  but as I said the 2nd topic "how can we prevent people who come from payment gateway at same time to not both pay" cannot be solved easily and globally (the solution discussed above would only work for some payment gateway) so we can close it safely.

So back to initial topic :

> If I have 2 products "A" in stock, add these 2 "A" products to cart, change the quantity to 1 for "A" in BO and click on "Confirm with an obligation to pay" on the last checkout step, I am able to pay while I should not. I should have at the last checkout step when I click on "Confirm with an obligation to pay" the info that there is only one product left and that the order cannot be validated. We do it already at the others checkout steps, we could do it also at the last step when the user clicks on "Confirm with an obligation to pay"

I suggest, because this issue strayed away from this topic, to re-create a new issue with only this single scenario.

**Comment by marionf** [comment]
@matks 

I think the confusion comes from the description of the issue that wasn't corresponding to the steps to reproduce.
So, I updated the title and the description and by the way, I specified the wanted behavior to fix this issue :)

**Comment by madsoliver** [comment]
I can confirm that this is still an issue with 1.7.7.5.

> What if 2 customers come back from the payment gateway in the same millisecond? 

_Millisecond_? No, you can open a checkout window monday and finish it friday. As long as the customer does **no changes on the cart, address or similar**, you can proceed with the order even though  the products in your cart are oversold. You can even open a checkout, then the **admin** sets the stock to 0 manually, well you are still able to proceed with the order. All my tests are with `ps_wirepayment`, so there are no delays from an external party. 

However, if the customer tries to change a product in the cart (does not have to be the oversold product), an ajax-call is made and then the customer gets a message that he/she cannot order the oos product. This check should be done before the order is created.

**Comment by Prestaworks** [comment]
> 
> 
> I can confirm that this is still an issue with 1.7.7.5.
> 
> > What if 2 customers come back from the payment gateway in the same millisecond?
> 
> _Millisecond_? No, you can open a checkout window monday and finish it friday. As long as the customer does **no changes on the cart, address or similar**, you can proceed with the order even though the products in your cart are oversold. You can even open a checkout, then the **admin** sets the stock to 0 manually, well you are still able to proceed with the order. All my tests are with `ps_wirepayment`, so there are no delays from an external party.
> 
> However, if the customer tries to change a product in the cart (does not have to be the oversold product), an ajax-call is made and then the customer gets a message that he/she cannot order the oos product. This check should be done before the order is created.

In your case the issue is in ps_wirepayment, that module could have it's own stock check in the confirmation controller.

If you put an extra stock check inside the validateOrder function, you will end up getting lots of payments from external payment gateways but no orders in the shop. So all those modules would have to add some kind of automatic refund feature. And it will still not help agains the use case where two orders are made on the same second as both those stock checks would return true.

Maybe it could be solved by a myIsam table that locks the table while reading / updating the stock, but that would be real bad on the performance of the page.

**Comment by madsoliver** [comment]
> If you put an extra stock check inside the validateOrder function, you will end up getting lots of payments from external payment gateways but no orders in the shop. 

Yes, that would not be a good option.

I'd say that the best option is `Cart::checkQuantities()` **before** payment is even begun. For offline and redirected payment solutions, the check can be done before you get redirected, for embedded/iframe/checkout payment solutions I think there is always some callback-function that can ajax-query `Cart::checkQuantities()` and abort the payment if it returns `false`.

**Comment by hibatallahAouadni** [comment]
Thanks @Hlavtox for the PR :rocket: :tada:
